File: code/ch12/go/norm_code.go
// Exercise 11 (§12.6.2) — the inherited, undocumented norm_code, Go variant.
// A nil *string plays Python's None; the ValueError becomes an error return.
// Note: len/[:8] count bytes, not characters — identical for the ASCII codes
// this function sees, another behavior a characterization suite would pin.
// Run: go test (norm_code_test.go exercises this file)
package delivery

import (
	"fmt"
	"strings"
	"unicode"
)

func normCode(s *string, strict bool) (*string, error) {
	if s == nil {
		if strict {
			return nil, nil
		}
		return new(string), nil // ""
	}
	t := strings.ReplaceAll(strings.ToUpper(strings.TrimSpace(*s)), "-", "")
	if len(t) > 8 {
		t = t[:8]
	}
	if strict && !isAlnum(t) {
		return nil, fmt.Errorf("invalid code: %q", t)
	}
	return &t, nil
}

func isAlnum(s string) bool { // Python's isalnum: nonempty, all letters/digits
	for _, r := range s {
		if !unicode.IsLetter(r) && !unicode.IsDigit(r) {
			return false
		}
	}
	return s != ""
}

File: code/ch12/javascript/norm_code.js
// Exercise 11 (§12.6.2) — the inherited, undocumented norm_code, JavaScript variant.
// null plays Python's None; RangeError plays ValueError; the Unicode property
// regex plays str.isalnum (nonempty, all letters/digits).
// Run: node norm_code.js
const assert = require("node:assert/strict");

function normCode(s, strict = false) {
  if (s === null) return strict ? null : "";
  s = s.trim().toUpperCase().replaceAll("-", "");
  if (s.length > 8) s = s.slice(0, 8);
  if (strict && !/^[\p{L}\p{N}]+$/u.test(s)) throw new RangeError(s);
  return s || (strict ? null : "");
}

// characterization pins, matching the Python original's observed behavior
assert.equal(normCode(null), "");
assert.equal(normCode(null, true), null);
assert.equal(normCode(" ab-12 "), "AB12");
assert.equal(normCode("abcdefghij"), "ABCDEFGH");   // truncated to 8
assert.equal(normCode("a!b"), "A!B");               // lenient keeps the junk
assert.throws(() => normCode("a!b", true), { name: "RangeError", message: "A!B" });
assert.throws(() => normCode("", true), RangeError); // empty fails isalnum too
assert.equal(normCode("  "), "");
assert.equal(normCode("--------"), "");
console.log("all characterization pins hold");
